File: py-langgraph/src/api/routes/images.py
import asyncio
import random

import httpx
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, Field

from src.config.settings import settings

router = APIRouter()


class ImageGenerationRequest(BaseModel):
    prompt: str = Field(..., min_length=1, max_length=2000)


class ImageGenerationResponse(BaseModel):
    image_data_url: str


MAX_IMAGE_REQUEST_ATTEMPTS = 3


# 调用需要 multipart 表单的 FLUX.2 模型接口，并在网络异常时重试。
async def _request_with_retry(url: str, api_key: str, prompt: str) -> httpx.Response:
    last_error: httpx.HTTPError | None = None
    for attempt in range(1, MAX_IMAGE_REQUEST_ATTEMPTS + 1):
        try:
            async with httpx.AsyncClient(timeout=60) as client:
                return await client.post(
                    url,
                    headers={"Authorization": f"Bearer {api_key}"},
                    files={
                        "prompt": (None, prompt),
                        "width": (None, "1024"),
                        "height": (None, "768"),
                    },
                )
        except httpx.HTTPError as exc:
            last_error = exc
            if attempt < MAX_IMAGE_REQUEST_ATTEMPTS:
                await asyncio.sleep(attempt + random.uniform(0, 1))

    raise last_error or RuntimeError("Image generation failed after retries")


# 从 Cloudflare 响应中提取可安全返回给调用方的错误信息。
def _get_error_message(payload: dict) -> str:
    errors = payload.get("errors") or []
    if errors and isinstance(errors[0], dict) and errors[0].get("message"):
        return str(errors[0]["message"])
    error = payload.get("error") or {}
    if isinstance(error, dict) and error.get("message"):
        return str(error["message"])
    return "Image generation failed"


@router.post("/generations", response_model=ImageGenerationResponse)
# 生成图片并将 Cloudflare 的 Base64 图片转为 data URL。
async def generate_image(request: ImageGenerationRequest):
    if not settings.image_api_key or not settings.image_base_url:
        raise HTTPException(status_code=503, detail="Image model is not configured")

    try:
        response = await _request_with_retry(
            f"{settings.image_base_url.rstrip('/')}/{settings.image_model}",
            settings.image_api_key,
            request.prompt,
        )
    except httpx.HTTPError:
        raise HTTPException(status_code=502, detail="Unable to reach the image service")

    try:
        payload = response.json() if response.content else {}
    except ValueError:
        payload = {}

    if response.is_error:
        raise HTTPException(status_code=response.status_code, detail=_get_error_message(payload))

    image_base64 = payload.get("result", {}).get("image")
    if not image_base64:
        raise HTTPException(status_code=502, detail=_get_error_message(payload))

    return ImageGenerationResponse(image_data_url=f"data:image/jpeg;base64,{image_base64}")

File: ts-langchain/src/api/routes/images.ts
import type { FastifyInstance } from "fastify";

import { config } from "../../config/index.js";

interface CloudflareImageResponse {
  result?: { image?: string };
  errors?: Array<{ message?: string }>;
  error?: { message?: string };
}

const MAX_IMAGE_REQUEST_ATTEMPTS = 3;

// 异步等待指定时长，用于请求重试的退避。
function wait(milliseconds: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

// 验证并拼接 Cloudflare Workers AI 模型运行接口地址。
function getImageApiUrl(): string | null {
  try {
    const url = new URL(config.IMAGE_BASE_URL);
    if (url.protocol !== "https:") return null;
    return `${url.toString().replace(/\/$/, "")}/${config.IMAGE_MODEL}`;
  } catch {
    return null;
  }
}

// 调用需要 multipart 表单的 FLUX.2 模型接口，并在网络异常时重试。
async function requestImageGeneration(
  url: string,
  apiKey: string,
  prompt: string,
): Promise<Response> {
  let lastError: unknown;

  for (let attempt = 1; attempt <= MAX_IMAGE_REQUEST_ATTEMPTS; attempt++) {
    try {
      const form = new FormData();
      form.append("prompt", prompt);
      form.append("width", "1024");
      form.append("height", "768");

      return await fetch(url, {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
        },
        body: form,
      });
    } catch (error) {
      lastError = error;
      if (attempt < MAX_IMAGE_REQUEST_ATTEMPTS) await wait(attempt * 1000);
    }
  }

  throw lastError;
}

// 从 Cloudflare 响应中提取可安全返回给调用方的错误信息。
function getErrorMessage(payload: CloudflareImageResponse | null): string {
  return (
    payload?.errors?.[0]?.message ||
    payload?.error?.message ||
    "Image generation failed"
  );
}

// 注册文本生图接口，并将 Cloudflare 的 Base64 图片转为 data URL。
export async function registerImageRoutes(app: FastifyInstance) {
  app.post<{ Body: { prompt?: string } }>(
    "/images/generations",
    async (request, reply) => {
      const prompt = request.body?.prompt?.trim();
      if (!prompt || prompt.length > 2000) {
        return reply
          .status(400)
          .send({ error: "prompt must be between 1 and 2000 characters" });
      }

      const imageApiUrl = getImageApiUrl();
      if (!imageApiUrl || !config.IMAGE_API_KEY) {
        return reply
          .status(503)
          .send({ error: "Image model is not configured" });
      }

      try {
        const response = await requestImageGeneration(
          imageApiUrl,
          config.IMAGE_API_KEY,
          prompt,
        );
        const payload = (await response
          .json()
          .catch(() => null)) as CloudflareImageResponse | null;

        if (!response.ok) {
          return reply.status(response.status).send({ error: getErrorMessage(payload) });
        }

        const imageBase64 = payload?.result?.image;
        if (!imageBase64) {
          return reply.status(502).send({ error: getErrorMessage(payload) });
        }

        return { image_data_url: `data:image/jpeg;base64,${imageBase64}` };
      } catch (error) {
        request.log.error(error, "Image generation request failed");
        return reply
          .status(502)
          .send({ error: "Unable to reach the image service" });
      }
    },
  );
}
